### ISSUE TITLE
Implement API caching using browser's cache storage

### ISSUE THREAD

**Issue by Yonghwan-Song** [Yonghwan-Song/pomodoro#15]
# Sequence Diagram

**The number of HTTP requests for the statistics can be significantly decreased by persisting data not only to the MongoDB but also to the Indexed DB.**

  Cached statistics becomes stale every time a new pomo record is persisted to server.
The advantage of caching is reduced significantly because what this app does is basically recording pomo data.
  To prevent this, I need to make the `useFetch()` persist data not only to the server but also to the `cache storage` to make the cached statistics not stale soon.

![image](https://github.com/Yonghwan-Song/pomodoro/assets/72689705/fc01d9ce-5788-4f2a-840b-4db9259ee146)

**Exception of case 1:** 

  in Settings.tsx

1. `createDemoData()` - cached statistics response is deleted. Because we cannot know yet what demo data will be at this point, we cannot cache the Demo data here unlike recording a pomo.
2. `removeDemoData()` - delete the cached statistics response. In other words, reset it.
3. `deleteAccount()` - delete the named cache

# Code

## 1.

**PatternTimer.tsx**

```tsx
async function recordPomo(user: User, duration: number, startTime: number) {
  try {
    const today = new Date(startTime);
    let LocaleDateString = `${
      today.getMonth() + 1
    }/${today.getDate()}/${today.getFullYear()}`;
    const idToken = await user.getIdToken();
    console.log(
      JSON.stringify({
        userEmail: user.email,
        duration,
        startTime,
        LocaleDateString,
      })
    );

    // update
    let cache = DynamicCache || (await openCache(CONSTANTS.CacheName));
    let statResponse = await cache.match(
      CONSTANTS.URLs.POMO + "/stat/" + user.email
    );
    if (statResponse !== undefined) {
      let statData = await statResponse?.json();
      console.log("statistics - ", statData);
      statData.push({
        userEmail: user.email,
        duration,
        startTime,
        date: LocaleDateString,
        isDummy: false,
      });
      cache.put(
        CONSTANTS.URLs.POMO + "/stat/" + user.email,
        new Response(JSON.stringify(statData))
      );
    }

    const response = await axios.post(
      CONSTANTS.URLs.POMO,
      {
        userEmail: user.email,
        duration,
        startTime,
        LocaleDateString,
      },
      {
        headers: {
          Authorization: "Bearer " + idToken,
        },
      }
    );
    console.log("res obj", response);
  } catch (err) {
    console.log(err);
  }
}
```

## 2.

**useFetch()**
```tsx
type DataType<T, S> = S extends undefined ? T : S;
type ArgType<T, S> = {
  urlSegment: string;
  modifier?: (arg: T) => DataType<T, S>;
  callbacks?: ((arg: DataType<T, S>) => void)[];
  additionalDeps?: DependencyList;
  additionalCondition?: boolean;
};
type CustomReturnType<T, S> = [
  DataType<T, S> | null,
  Dispatch<SetStateAction<DataType<T, S> | null>>
];

export function useFetch<T, S = undefined>({
  urlSegment,
  modifier,
  callbacks,
  additionalDeps,
  additionalCondition,
}: ArgType<T, S>): CustomReturnType<T, S> {
  const [data, setData] = useState<DataType<T, S> | null>(null);
  const { user } = UserAuth()!;

  let moreDeps: DependencyList = additionalDeps ?? [];

  useEffect(() => {
    async function fetchData() {
      try {
        const idToken = await user?.getIdToken();
        const response = await axios.get(urlSegment + `/${user!.email}`, {
          headers: {
            Authorization: "Bearer " + idToken,
          },
        });

        let data =
          modifier !== undefined ? modifier(response.data as T) : response.data;

        setData(data);
        if (callbacks !== undefined) {
          callbacks.forEach((fn) => {
            fn(data);
          });
        }

        return response;
      } catch (error) {
        console.warn(error);
      }
    }

    async function getData() {
      let resData = await caches.match(urlSegment + `/${user!.email}`);
      if (resData !== undefined) {
        let data =
          modifier !== undefined
            ? modifier((await resData.json()) as T)
            : await resData.json();

        setData(data);
        if (callbacks !== undefined) {
          callbacks.forEach((fn) => {
            fn(data);
          });
        }
      } else {
        let res = await fetchData();
        let resFetched = new Response(JSON.stringify(res?.data));
        DynamicCache?.put(urlSegment + `/${user!.email}`, resFetched);
      }
    }

    if (user !== null && (additionalCondition ?? true)) {
      getData();
    }

    console.log("Fetching Data", typeof data);
  }, [user, ...moreDeps]);

  return [data, setData];
}
```

**Comment by Yonghwan-Song** [comment]
Completed by the commit 9444872.

**Comment by Yonghwan-Song** [comment]
Solved by the 9444872.
